File: ruby/easy/count_binary_substrings.rb
# @param {String} s
# @return {Integer}
def count_binary_substrings(s)
    return 0 unless s
    i, count, prev_count, total_count = 1, 1, 0, 0
    c = s[0]
    while i < s.length
        if(s[i] == c)
            count += 1 and i += 1
            next
        end
        total_count += [count, prev_count].min if(prev_count > 0)
        prev_count = count
        c = s[i]
        count = 1
        i += 1
    end
    return total_count + [count, prev_count].min
end

File: javascript/easy/countBinarySubstrings.js
/**
 * @param {string} s
 * @return {number}
 */
var countBinarySubstrings = function(s) {
    if(s.length === 0) return 0;
    let count = 1, totalCount = 0, prevCount = 0, c = s[0], i = 1;
    while(i < s.length) {
        if(s[i] === c) {
            count++;
            i++; continue;
        }
        if(prevCount > 0) {
            totalCount += Math.min(prevCount, count);
        }
        c = s[i];
        prevCount = count;
        count = 1;
        i++;
    }
    return totalCount + Math.min(prevCount, count);
};
